cub/docs/developer_overview.rst

```diff
@@ -744,3 +744,198 @@ contains hidden visibility attribute.
+We reserve the right to select the implementation with stronger guarantees compared to what user requested.
+The remaining part of this section is only related to CUB developers.
+
+
+Guarantees
+====================================
+
+The best form of documentation is code.
+The default guarantees should be visible to the user in a convenient way outside of the documentation.
+This is achieved as follows:
+
+
+.. code-block:: c++
+
+   struct reduce
+   {
+      using default_guarantees_t =                                                                                          // (1)
+        cub::detail::guarantees::guarantees_t<cub::detail::guarantees::determinism_not_guaranteed_t,
+                                              detail::max_memory_footprint_t>;
+
+      template <class IteratorT, class ReguirementsT = default_guarantees_t>
+      static auto sum(IteratorT begin, IteratorT end, ReguirementsT requirements = default_guarantees_t())                  // (2)
+      {
+        auto guarantees    = cub::detail::requirements::mask(default_guarantees_t(), requirements);                         // (3)
+        auto max_footprint = cub::detail::requirements::get<detail::max_memory_footprint_t>(guarantees);                    // (4)
+        auto determinism = cub::detail::requirements::get<cub::detail::guarantees::run_to_run_deterministic_t>(guarantees); // (5)
+        return reduce::sum(begin, end, determinism, max_footprint);                                                         // (6)
+      }
+   };
+
+
+Above, the default guarantees are provided in (1). 
+CUB API receives requirements in (2).
+User-specified requirements mask the default guarantees in (3).
+Concrete guarantees are queried by providing category representative in (4) and (5).
+Finally, tag dispatch can be used to select the proper implementation in (6).
+
+Tag dispatch is used for simple cases.
+Given many requirement categories, it might lead to some complications.
+To address this issue, algorithm author might approach dispatch differently.
+
+Since each requirement is ordered within the category,
+it's possible to order implementation by strength.
+
+
+.. code-block:: c++
+
+    class scan
+    {
+      using weak_guarantees_t = cub::detail::guarantees::guarantees_t<cub::detail::guarantees::determinism_not_guaranteed_t,
+                                                                      detail::discard_partials_t>;
+
+      using strong_guarantees_t = cub::detail::guarantees::guarantees_t<cub::detail::guarantees::run_to_run_deterministic_t,
+                                                                        detail::preserve_partials_in_t<void>>;
+
+      using default_guarantees_t = weak_guarantees_t;
+
+      template <class IteratorT, class ReguirementsT>
+      static typename ::cuda::std::enable_if<
+        cub::detail::guarantees::statically_satisfy_t<weak_guarantees_t, ReguirementsT>::value,             // (1)
```

Comment:
Typo:

```suggestion
        cub::detail::guarantees::statically_satisfy_t<weak_guarantees_t, RequirementsT>::value,             // (1)
```

Reply:
Question: How does this approach interact with non-static requirements, e.g. memory footprint?